pkg/util/ec2/ec2_tags.go

```diff
@@ -29,6 +30,46 @@ var (
 )
 
 func fetchEc2Tags(ctx context.Context) ([]string, error) {
+	if config.Datadog.GetBool("collect_ec2_tags_use_imds") {
+		// prefer to fetch tags from IMDS, falling back to the API
+		tags, err := fetchEc2TagsFromIMDS(ctx)
+		if err == nil {
+			return tags, nil
+		}
+
+		log.Debugf("Could not fetech tags from instance metadata (trying EC2 API instead): %s", err)
+	}
+
+	tags, err := fetchEc2TagsFromAPI(ctx)
+	if err == nil {
+		return tags, err
+	}
+
+	return nil, err
+}
+
+func fetchEc2TagsFromIMDS(ctx context.Context) ([]string, error) {
+	keysStr, err := getMetadataItem(ctx, "/tags/instance")
+	if err != nil {
+		return nil, err
+	}
+
+	// keysStr is a newline-separated list of strings containing tag keys
+	keys := strings.Split(keysStr, "\n")
+
+	tags := []string{}
+	for _, key := range keys {
+		val, err := getMetadataItem(ctx, "/tags/instance/"+key)
```

Comment:
Are tag keys always a valid url component? Would it make sense to escape them here?

Reply:
> If you enable instance tags in instance metadata, instance tag keys can only use letters (a-z, A-Z), numbers (0-9), and the following characters: -_+=,.@:. Instance tag keys can't use spaces, /, or the reserved names ., .., or _index.

I'll add a comment.